include/ibamr/BrinkmanPenalizationRigidBodyDynamics.h

```diff
@@ -0,0 +1,372 @@
+ * motion.
+ *
+ * For further information on applications of this class see
+ * https://arxiv.org/abs/1904.04078.
+ */
+class BrinkmanPenalizationRigidBodyDynamics : public BrinkmanPenalizationStrategy
+{
+    ////////////////////////////// PUBLIC ////////////////////////////////////////
+
+public:
+    /*
+     * \brief Constructor of the class.
+     */
+    BrinkmanPenalizationRigidBodyDynamics(std::string object_name,
+                                          SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_solid_var,
+                                          SAMRAI::tbox::Pointer<IBAMR::AdvDiffHierarchyIntegrator> adv_diff_solver,
+                                          SAMRAI::tbox::Pointer<IBAMR::INSVCStaggeredHierarchyIntegrator> fluid_solver,
+                                          SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db = nullptr,
+                                          bool register_for_restart = true);
+
+    /*
+     * \brief Destructor of the class.
+     */
+    ~BrinkmanPenalizationRigidBodyDynamics();
+
+    /*!
+     * \brief Set initial conditions for the rigid body.
+     *
+     * \note It is OK to set some or all of these quantities to zero, because the fully-prescribed
+     * kinematics case does not need any information of mass and moment of inertia.
+     */
+    void setInitialConditions(const Eigen::Vector3d& X_com,
+                              const Eigen::Vector3d& U_com = Eigen::Vector3d::Zero(),
+                              const Eigen::Vector3d& W_com = Eigen::Vector3d::Zero(),
+                              const double& mass = 0.0,
+                              const Eigen::Matrix3d& J_com = Eigen::Matrix3d::Zero(),
+                              const Eigen::Quaterniond& quaternion = Eigen::Quaterniond::Identity());
+
+    /*!
+     * \brief Typedef specifying interface for specifying rigid body velocities.
+     */
+    using KinematicsFcnPtr = void (*)(double data_time, Eigen::Vector3d& U_com, Eigen::Vector3d& W_com, void* ctx);
+
+    /*
+     * \brief Kinematics function data.
+     */
+    struct KinematicsFcnData
+    {
+        KinematicsFcnData(KinematicsFcnPtr comvelfcn = nullptr, void* ctx = nullptr) : comvelfcn(comvelfcn), ctx(ctx)
+        {
+            // intentionally left blank
+        }
+        KinematicsFcnPtr comvelfcn;
+        void* ctx;
+    };
+
+    /*!
+     * \brief Register user defined constrained velocity functions.
+     */
+    void registerKinematicsFunction(KinematicsFcnPtr comvelfcn, void* ctx = nullptr);
+
+    /*!
+     * \brief Register user defined velocity function data.
+     */
+    void registerKinematicsFunction(const KinematicsFcnData& data);
+
+    /*!
+     * \brief Set what rigid DOFs need to be solved for this
+     * particular structure.
+     */
+    virtual void setSolveRigidBodyVelocity(const IBTK::FreeRigidDOFVector& solve_rigid_dofs);
```

Comment:
At some point, should we make a common RBD base class to avoid reimplementing this?